### PR TITLE
Refactor #68: Database Catalog TTL

### PR DESCRIPTION
**Pull request by vincentsarago**
This PR updates #68:

- move settings to the API settings 
- move state variable to the `Database` object (`last_updated`) 
- ~~add `lastUpdated` in the `/collections` response (NOT IN THE SPECIFICATION)~~
- remove the `refresh` catch in the middleware

### CODE REVIEW

File: tipg/settings.py [hunk 1]
```diff
@@ -104,6 +103,8 @@ class APISettings(pydantic.BaseSettings):
 
     add_tiles_viewer: bool = True
 
+    catalog_ttl: int = 300
```
Comment (vincentsarago) [inline tipg/settings.py:106]: Change the `catalog_ttl` to be a string so we can easily set it with env variable 

```
TIPG_CATALOG_TTL=10 uvicorn tipg.main:app --port 8080 --reload
```

File: tipg/middleware.py [hunk 2]
```diff
@@ -51,19 +51,38 @@ async def dispatch(self, request: Request, call_next):
...
-            await register_collection_catalog(request.app)
-            response = await call_next(request)
-        elif datetime.now() > request.app.state.catalog_expire:
+        collection_catalog = getattr(request.app.state, "collection_catalog", {})
+        collections = collection_catalog.get("collections")
+        last_updated = collection_catalog.get("last_updated")
+
+        if not collections or datetime.now() > (
+            last_updated + timedelta(seconds=self.ttl)
+        ):
```
Comment (vincentsarago) [inline tipg/middleware.py:77]: 2 cases:
- there is not collections: we update (🤷‍♂️) 
- collections are too old: we update

File: tipg/middleware.py [hunk 3]
```diff
@@ -51,19 +51,38 @@ async def dispatch(self, request: Request, call_next):
 class CatalogUpdateMiddleware(BaseHTTPMiddleware):
     """Middleware to update the catalog cache."""
 
+    def __init__(
+        self,
+        app: ASGIApp,
+        ttl: int = 300,
```
Comment (vincentsarago) [inline tipg/middleware.py:57]: we could maybe add a special case when `ttl=0` meaning that we don't want any refresh 🤷‍♂️

Comment (vincentsarago) [reply tipg/middleware.py:57]: that could be handled at the application level
```python
if settings.catalog_ttl:
    app.add_middleware(CatalogUpdateMiddleware, ttl=settings.catalog_ttl)

**Comment by vincentsarago** [comment]
tests are failing for some weird rounding issue in python 3.11. Not related to any change in this PR

File: tipg/factory.py [hunk 4]
```diff
@@ -527,6 +532,7 @@ def collections(
                 links=links,
                 numberMatched=matched_items,
                 numberReturned=items_returned,
+                lastUpdated=str(collection_catalog["last_updated"]),
```
Comment (bitner) [inline tipg/factory.py:535]: let's just leave this out since it's not part of the spec

Comment (vincentsarago) [reply tipg/factory.py:535]: 👍

File: tipg/middleware.py [hunk 5]
```diff
@@ -51,19 +51,38 @@ async def dispatch(self, request: Request, call_next):
 class CatalogUpdateMiddleware(BaseHTTPMiddleware):
```
Comment (bitner) [inline tipg/middleware.py:51]: I thought the biggest thing you wanted to do was to get rid of BaseHTTPMiddleware???

Comment (vincentsarago) [reply tipg/middleware.py:51]: I will in an other PR

**Review by bitner — APPROVED** [review]
I'm cool with all these changes. I thought the biggest thing you wanted to do though was to get rid of BaseHTTPMiddleware though???

File: tipg/middleware.py (hunk 2 above)
Comment (vincentsarago) [reply tipg/middleware.py:77]: changed this to only rely on the `last_updated` tag. 

Starting the application with a brand new database could in theory mean that every request would try to refresh the catalog, which IMO seems a bit risky

Comment (vincentsarago) [reply tipg/middleware.py:79]: yeah but what if the `startup` fails or if there is not collections at all 🤷 

I think it's fine as is right now!

**Pull request merged**